RecFCCeeCalorimeter/src/components/AugmentClustersFCCee.cpp

```diff
@@ -0,0 +1,253 @@
+#include "AugmentClustersFCCee.h"
+
+// k4FWCore
+#include "k4Interface/IGeoSvc.h"
+
+// edm4hep
+#include "edm4hep/ClusterCollection.h"
+
+// DD4hep
+#include "DD4hep/Detector.h"
+
+// ROOT
+#include "TLorentzVector.h"
+#include "TString.h"
+#include "TVector3.h"
+#include "TMath.h"
+
+DECLARE_COMPONENT(AugmentClustersFCCee)
+
+AugmentClustersFCCee::AugmentClustersFCCee(const std::string &name, ISvcLocator *svcLoc)
+    : Gaudi::Algorithm(name, svcLoc),
+      m_geoSvc("GeoSvc", "AugmentClustersFCCee")
+{
+  declareProperty("inClusters", m_inClusters, "Input clusters");
+  declareProperty("outClusters", m_outClusters, "Output clusters");
+}
+
+StatusCode AugmentClustersFCCee::initialize()
+{
+  {
+    StatusCode sc = Gaudi::Algorithm::initialize();
+    if (sc.isFailure())
+      return sc;
+  }
+
+  // check if readouts exist
+  for (size_t k = 0; k < m_readoutNames.size(); k++)
+  {
+    std::string readoutName = m_readoutNames[k];
+    if (m_geoSvc->getDetector()->readouts().find(readoutName) == m_geoSvc->getDetector()->readouts().end())
+    {
+      error() << "Readout <<" << readoutName << ">> does not exist." << endmsg;
+      return StatusCode::FAILURE;
+    }
+  }
+
+  // Check if readout related variables have the same size
+  if (m_systemIDs.size() != m_readoutNames.size())
+  {
+    error() << "Sizes of the systemIDs vector and readoutNames vector do not match, exiting!" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  if (m_systemIDs.size() != m_numLayers.size())
+  {
+    error() << "Sizes of systemIDs vector and numLayers vector do not match, exiting!" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  if (m_systemIDs.size() != m_layerFieldNames.size())
+  {
+    error() << "Sizes of systemIDs vector and layerFieldNames vector do not match, exiting!" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  if (m_systemIDs.size() != m_detectorNames.size())
+  {
+    error() << "Sizes of systemIDs vector and detectorNames vector do not match, exiting!" << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  // initialise the list of metadata for the clusters
+  // append to the metadata of the input clusters
+  std::string showerShapeDecorations = m_inShapeParameterHandle.get();
+  if (showerShapeDecorations != "")
+    showerShapeDecorations += ":";
+  for (size_t k = 0; k < m_detectorNames.size(); k++)
+  {
+    const char *detector = m_detectorNames[k].c_str();
+    for (unsigned layer = 0; layer < m_numLayers[k]; layer++)
+    {
+      if (layer > 0)
+        showerShapeDecorations += ":";
+      showerShapeDecorations += TString::Form("f%s%d:theta%s%d:phi%s%d",
```

Comment:
`TString::Format()` should work :)

Reply:
Hi @kjvbrt I haven't tested this change myself, but the original Form(..) method in global namespace was returning a  const char*, which we can then add to showerShapeDecorations (which is a std::string).
Not sure that with TString::Form or TString::Format (that both return a TString) the code will still compile. @AlexisMlz did you try to compile the code?

See https://root.cern.ch/doc/master/classTString.html#aeb44d6183d8b1f1b7090dbd3c93f5e39

Note: this is not to be confused with [Format](https://root.cern.ch/doc/master/classTString.html#ac34bbb89627e20624a416dfbb7a3db70) and [Form](https://root.cern.ch/doc/master/classTString.html#aeb44d6183d8b1f1b7090dbd3c93f5e39) (in the global namespace) which returns a const char* and relies on a thread-local static character buffer.

Reply:
BTW what was the problem with using Form(..) in first place ? 
https://root.cern.ch/doc/master/TString_8cxx_source.html#l02489